lenses-annotation-processor/src/jvmMain/kotlin/dev/fritz2/lens/LensesProcessor.kt

```diff
@@ -28,169 +26,650 @@ import dev.fritz2.core.Lenses
+                    classDeclaration.toClassName(),
+                    attributeName
+                )
+                .build()
+        )
+    }
+
+    private val createLens: FunSpec.Builder.(MemberName, KSClassDeclaration) -> FunSpec.Builder = { attributeName, _ ->
+        addCode(
+            """ 
+        |return %M(
+        |    "%L",
+        |    { it.%M },
+        |    { p, v -> p.copy(%M = v)}
+        |  )
+        """.trimMargin(),
+            MemberName("dev.fritz2.core", "lensOf"),
+            attributeName,
+            attributeName,
+            attributeName
+        )
+    }
+
+    private val createDelegatingLens: FunSpec.Builder.(MemberName, KSClassDeclaration) -> FunSpec.Builder =
+        { attributeName, classDeclaration ->
+            val children = classDeclaration.getSealedSubclasses()
+
+            addCode(
+                """ 
+            |return %M(
+            |    "%L",
+            |    { parent ->
+            |        when(parent) {
+            """.trimMargin(),
+                MemberName("dev.fritz2.core", "lensOf"),
+                attributeName
+            )
+            addStatement("")
+            children.forEach { child ->
+                addStatement(
+                    """ 
+                |            is %T -> parent.%M
+                """.trimMargin(),
+                    child.toClassName(),
+                    attributeName
                 )
             }
-        }.build()
+            addCode(
+                """
+            |        }
+            |    },
+            |    { parent, value ->
+            |        when(parent) {
+            """.trimMargin(),
+            )
+            addStatement("")
+            children.forEach { child ->
+                addStatement(
+                    """ 
+                |            is %T -> parent.copy(%M = value)
+                """.trimMargin(),
+                    child.toClassName(),
+                    attributeName
+                )
+            }
+            addCode(
+                """
+            |        }
+            |    }
+            |)
+            """.trimMargin(),
+            )
+        }
+}
 
-        fileSpec.writeTo(codeGenerator = codeGenerator, aggregating = false)
+private fun String.lowerCamelCased() = "${first().lowercase()}${drop(1)}"
+
+private enum class TypeVariant {
+    Invalid, DataClass, SealedDataClass, SealedInterface
+}
+
+private fun KSClassDeclaration.isTypeVariant(): TypeVariant = when (classKind) {
+    ClassKind.INTERFACE ->
+        if (modifiers.contains(Modifier.SEALED)) TypeVariant.SealedInterface
+        else TypeVariant.Invalid
+
+    ClassKind.CLASS ->
+        if (modifiers.contains(Modifier.DATA)) TypeVariant.DataClass
+        else if (modifiers.contains(Modifier.SEALED)) TypeVariant.SealedDataClass
+        else TypeVariant.Invalid
+
+    else -> TypeVariant.Invalid
+}
+
+private fun interface DetermineLenseableProperties {
```

Comment:
Maybe this could be `LenseablePropertiesDeterminer` instead? This way the name would be a noun